src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,977 @@
+                foreach (var identifierName in queryExpressionProcessingInfo.IdentifierNames)
+                {
+                    // Identifier from the foreach can already be in scope of the foreach statement.
+                    if (forEachStatement.Identifier.ValueText != identifierName)
+                    {
+                        if (_semanticFacts.GenerateUniqueName(
+                                _semanticModel,
+                                location: forEachStatement.Statement,
+                                containerOpt: forEachStatement.Statement,
+                                baseName: identifierName,
+                                usedNames: Enumerable.Empty<string>(),
+                                _cancellationToken).ValueText != identifierName)
+                        {
+                            documentUpdateInfo = default;
+                            return false;
+                        }
+                    }
+                }
+
+                // If query does not contains identifier with the same name as declared in the foreach,
+                // declare this identifier in the body.
+                if (!queryExpressionProcessingInfo.ContainsIdentifier(forEachStatement.Identifier))
+                {
+                    documentUpdateInfo = ConvertIfInToForeachWithExtraVariableDeclaration(forEachStatement, queryExpressionProcessingInfo);
+                    return true;
+                }
+                else
+                {
+                    // The last select expression in the query returns this identifier:
+                    // foreach(var thisIdentifier in from ....... select thisIdentifier)
+                    // if thisIdentifier in foreach is var, it is OK
+                    // if a type is specified for thisIdentifier in forEach, check that the type is the same as in the select expression
+                    // foreach(MyType thisIdentifier in from ....... from MyType thisIdentifier ... select thisIdentifier)
+
+                    // The last clause in query stack must be SelectClauseSyntax.
+                    var lastSelectExpression = ((SelectClauseSyntax)queryExpressionProcessingInfo.Stack.Peek()).Expression;
+                    if (lastSelectExpression is IdentifierNameSyntax identifierName)
+                    {
+                        if (forEachStatement.Identifier.ValueText == identifierName.Identifier.ValueText)
+                        {
+                            var forEachStatementTypeSymbolType = _semanticModel.GetTypeInfo(forEachStatement.Type, _cancellationToken).Type;
+                            var lastSelectExpressionTypeInfo = _semanticModel.GetTypeInfo(lastSelectExpression, _cancellationToken);
+                            if (SymbolEquivalenceComparer.Instance.Equals(lastSelectExpressionTypeInfo.ConvertedType, lastSelectExpressionTypeInfo.Type) &&
```

Comment:
>lastSelectExpressionTypeInfo.Type [](start = 118, length = 33)

I guess we can allow ```lastSelectExpressionTypeInfo.Type``` to be null. To handle cases like ```... select null```

Reply:
do you have an example for this? should we have a conversion to null?

---
In reply to: [178423255](https://github.com/dotnet/roslyn/pull/25362#discussion_r178423255) [](ancestors = 178423255)

Reply:
> do you have an example for this? should we have a conversion to null?

I can have a Select operator with a delegate that transforms int to string and a query like ```from a in new [] {1,2,3} select null``` 
Type will be null
Converted type will be string and we should still require it to match the forEachStatementTypeSymbolType

ConvertedType should also match Type, unless the Type is null 
 


---
In reply to: [178423496](https://github.com/dotnet/roslyn/pull/25362#discussion_r178423496) [](ancestors = 178423496,178423255)

Reply:
Regarding

>ConvertedType should also match Type, unless the Type is null 

It looks like we are getting here only when the last select expression is an identifier. Its type will never be null. No need to worry about this situation. Sorry for the confusion, long day. 

---
In reply to: [178423768](https://github.com/dotnet/roslyn/pull/25362#discussion_r178423768) [](ancestors = 178423768,178423496,178423255)